### PR TITLE
Sparsity pattern detection

### PR DESCRIPTION
**Pull request by michel2323**
This enables sparisty detection of the Jacobians. This takes place on the CPU. Unfortunately I couldn't get the KA residual to work with the SparsityDetection.jl.

The coloring etc now takes place in the AD module and is not in solve() anymore.

### CODE REVIEW

File: src/ExaPF.jl [hunk 1]
```diff
@@ -107,6 +94,37 @@ function residualFunction_real!(F, v_re, v_im,
 
     return F
 end
+function residualFunction_polar_sparsity!(F, v_m, v_a,
```
[the remark this answers is not in the document]
Comment (frapac) [reply src/ExaPF.jl:97]: Also, if we are using this function only in the sparsity detection pass, do you know if we could use the function `residualFunction` below instead?

File: src/ExaPF.jl [hunk 2]
```diff
@@ -142,8 +160,8 @@ end
 end
 
 function residualFunction_polar!(F, v_m, v_a,
-                     ybus_re_nzval, ybus_re_colptr, ybus_re_rowval,
-                     ybus_im_nzval, ybus_im_colptr, ybus_im_rowval,
+                     ybus_re,
```
[the remark this answers is not in the document]
Comment (dmaldona) [reply src/ExaPF.jl:163]: Yes, very nice touch.

File: src/helpers.jl [hunk 3]
```diff
@@ -0,0 +1,16 @@
+mutable struct Spmat{T}
```
[the remark this answers is not in the document]
Comment (michel2323) [reply src/helpers.jl:1]: It's also something we should think about. There might be a smarter way to wrap `CuSparseMatrixCSR` and `SparseMatrixCSC`.

File: src/ExaPF.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (michel2323) [reply src/ExaPF.jl:97]: Above? I wanted it to be the same code then the KA one. The others don't seem to fit. We could actually start removing those.

File: src/algorithms/precondition.jl [hunk 4]
```diff
@@ -28,6 +28,11 @@ mutable struct Preconditioner <: AbstractPreconditioner
     cupart
     P
     function Preconditioner(J, npart, device=CPU())
+        if isa(J, CuSparseMatrixCSR)
+            J = SparseMatrixCSC(J)
+        # else
```
[the remark this answers is not in the document]
Comment (michel2323) [reply src/algorithms/precondition.jl]: Done

**Pull request merged**